File: 0310-minimum-height-trees/0310-minimum-height-trees.cpp
typedef vector<vector<int>> GRAPH;
class Solution
{
public:
    void addEdge(GRAPH &adjList, int from, int to)
    {
        adjList[from].push_back(to);
        adjList[to].push_back(from);
    }

    vector<int> topSort(GRAPH &adjList)
    {
        int sz = adjList.size();

        if (sz == 1) // special case
            return {0};
        else if (sz == 2)
            return {0, 1};
        queue<int> q;
        vector<int> deg(adjList.size(), 0);
        for (auto node : adjList)
        {
            for (int neigbour : node)
            {
                deg[neigbour]++;
            }
        }

        for (int i = 0; i < deg.size(); i++)
        {
            if (deg[i] == 1)
                q.push(i);
        }
        vector<int> res;
        for (int sz = q.size(); !q.empty(); sz = q.size())
        {
            res.clear();
            while (sz--)
            {
                int node = q.front();
                q.pop();
                res.push_back(node);
                for (int neighbour : adjList[node])
                    if (--deg[neighbour] == 1)
                        q.push(neighbour);
            }
        }

        return res;
    }
    vector<int> findMinHeightTrees(int n, vector<vector<int>> &edges)
    {
        GRAPH graph(edges.size() + 1);

        for (auto &edge : edges)
            addEdge(graph, edge[0], edge[1]);

        return topSort(graph);
    }
};

File: 0310-minimum-height-trees/0310-minimum-height-trees.java
import java.util.*;

class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {

        if (n==1)
            return List.of(0);
        HashMap<Integer, List<Integer>> graph = new HashMap<>();
        int[] inDeg = new int[n];
        Queue<Integer> queue = new LinkedList<>();
        List<Integer> res = new ArrayList<>();
        
        
        buildUndirectedGraph(edges, graph);
        buildInDegArray(n, graph, inDeg);
        startWithLeafNodes(n, inDeg, queue);
        res = bfs(graph, inDeg, queue, res);

        return res;
    }

    private static List<Integer> bfs(HashMap<Integer, List<Integer>> graph, int[] inDeg, Queue<Integer> queue, List<Integer> res) {
        for (int sz = queue.size(); !queue.isEmpty(); sz = queue.size()) {
            res = new ArrayList<>();
            while (sz-- != 0) {
                int node = queue.poll();
                res.add(node);
                for (int neighbour : graph.getOrDefault(node, new ArrayList<>())) {
                    if (--inDeg[neighbour] == 1) {
                        queue.add(neighbour);
                    }
                }
            }
        }
        return res;
    }

    private static void startWithLeafNodes(int n, int[] inDeg, Queue<Integer> queue) {
        for (int i = 0; i < n; i++) {
            if (inDeg[i] == 1) {
                queue.add(i);
            }

        }
    }

    private static void buildInDegArray(int n, HashMap<Integer, List<Integer>> graph, int[] inDeg) {
        for (int i = 0; i < n; i++) {
            for (int neighbour : graph.getOrDefault(i, new ArrayList<>())) {
                inDeg[neighbour]++;
            }
        }
    }

    private static void buildUndirectedGraph(int[][] edges, HashMap<Integer, List<Integer>> graph) {
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            graph.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
        }
    }
}